src/transport/SecureSessionTable.cpp

```diff
@@ -183,6 +220,100 @@ SecureSession * SecureSessionTable::EvictAndAllocate(uint16_t localSessionId, Se
+        //
+        // Sorting on Key1
+        //
+        if (a.mNumMatchingOnFabric > b.mNumMatchingOnFabric)
+        {
+            return true;
+        }
+
+        if (a.mNumMatchingOnFabric == b.mNumMatchingOnFabric)
+        {
+            int ScoreA = a.mSession->GetPeer().GetFabricIndex() == evictionContext.GetSessionEvictionHint().GetFabricIndex();
+            int ScoreB = b.mSession->GetPeer().GetFabricIndex() == evictionContext.GetSessionEvictionHint().GetFabricIndex();
+
+            //
+            // Sorting on Key2
+            //
+            if (ScoreA > ScoreB)
+            {
+                return true;
+            }
+
+            if (ScoreA == ScoreB)
+            {
+                //
+                // Sorting on Key3
+                //
+                if (a.mNumMatchingOnPeer > b.mNumMatchingOnPeer)
+                {
+                    return true;
+                }
+
+                if (a.mNumMatchingOnPeer == b.mNumMatchingOnPeer)
+                {
+                    int ScoreA1 = a.mSession->GetPeer() == evictionContext.GetSessionEvictionHint();
+                    int ScoreB1 = b.mSession->GetPeer() == evictionContext.GetSessionEvictionHint();
+
+                    //
+                    // Sorting on Key4
+                    //
+                    if (ScoreA1 > ScoreB1)
+                    {
+                        return true;
+                    }
+
+                    if (ScoreA1 == ScoreB1)
+                    {
+                        int aStateScore = 0, bStateScore = 0;
+
+                        auto assignStateScore = [](auto & score, const auto & session) {
+                            if (session.IsDefunct())
+                            {
+                                score = 2;
+                            }
+                            else if (session.IsActiveSession())
+                            {
+                                score = 1;
+                            }
+                            else
+                            {
+                                score = 0;
+                            }
+                        };
+
+                        assignStateScore(aStateScore, *a.mSession);
+                        assignStateScore(bStateScore, *b.mSession);
+
+                        //
+                        // Sorting on Key5
+                        //
+                        if (aStateScore > bStateScore)
+                        {
+                            return true;
+                        }
+
+                        if (aStateScore == bStateScore)
+                        {
+                            //
+                            // Sorting on Key6
+                            //
+                            return (a->GetLastActivityTime() < b->GetLastActivityTime());
+                        }
+                    }
+                }
+            }
+        }
+
+        return false;
+    });
+
+#if 0
```

Comment:
Which do folks prefer?

Reply:
Previous one, however needs indent/logic updates as I commented.
Multiple sorts feels slow/wasteful.

Also wondering if we just need "top" do we actually need a sort? Finding "min/max" is generally a faster operation than a sort (linear vs n lgn at best, plus more moving things overhead).

Reply:
Saw the comment regarding "cannot realy evict, must move to the next".
Do we expect in the future to be able to just min/max evict? if so, should there be a TODO and comment explaining why we sort instead of "select the session to evict" ?

Reply:
There are comments already about the speculative nature of eviction (i.e marking for eviction doesn't guarantee it frees up a slot), but it doesn't quite tie it to the need for sorting. So I can definitely add a comment there.

In the future, if we do ever get around to making `SessionHolder` a true weak ref (see #19495), then yes, we can always **guarantee** that the session we select will always be evictable, because we'd select sessions that have no strong-refs attached to them.

For now, that isn't possible. However, I think this raises a separate, interesting possibility - in the majority of cases, it is likely that the session selected for eviction will actually be evicted (i.e it shouldn't have strong refs attached to it), so just selecting the most likely candidate using an O(n) traversal, will suffice for most of those cases. In the cases where it doesn't, it starts to de-generate, but that shouldn't happen all that often.

For the purposes of getting this PR in, I'll add the comment, and file an issue exploring the O(n) linear min/max possibility.